text/0000-alloc-me-maybe.md

```diff
@@ -0,0 +1,547 @@
+    -> Result<(), (I::IntoIter, Option<T>, Err)>
+```
+
+Note that this API only even works because Iterator's signature currently guarantees that the yielded elements outlive the iterator. This would not be the case if we ever moved to support so-called "streaming iterators", which yield elements that point into themselves.
+
+
+
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Due to the diversity of requirements between our user profiles, there isn't any one-size fits all solution. This RFC proposes two solutions which will require minimal work for maximal impact:
+
+* For the server users, an `oom=panic` configuration, in the same vein as the `panic=abort`.
+* For everyone else, add `try_reserve` and `try_reserve_exact` as standard collection APIs.
+
+
+
+## oom=panic
+
+Applying this configuration in a Cargo.toml would change the behaviour of the global allocator's `oom()` function, which currently aborts, to instead panic. As discussed in the Server user profile, this would allow OOM to be handled at task boundaries with minimal effort for server developers, and no effort from library maintainers.
+
+If using a thread-per-task model, OOMs will be naturally caught at the thread boundary. If using a different model, tasks can be isolated using the `thread::catch_unwind` or `Future::catch_unwind` APIs.
+
+We expose a flag, rather than changing the default, because we maintain that *by default* Rust programmers should not be trying to recover from allocation failures.
+
+For instance, a project which desires to work this way would add this to their Cargo.toml:
+
+```toml
+[profile]
+oom = panic
+```
+
+And then in their application, do something like this:
+
+```rust
+fn main() {
+    set_up_event_queue();
+    loop {
+        let event = get_next_event();
+        let result = ::std::panic::catch_unwind(|| {
+            process_event(&mut event)
+        });
+
+        if let Err(err) = result {
+            if let Some(message) = err.downcast_ref::<&str>() {
+                eprintln!("Task crashed: {}", message);
+            } else if let Some(message) = err.downcast_ref::<String>() {
+                eprintln!("Task crashed: {}", message);
+            } else {
+                eprintln!("Task crashed (unknown cause)");
+            }
+
+            // Handle failure...
+        }
+    }
+}
+```
+
+
+
+
+## try_reserve
+
+`try_reserve` and `try_reserve_exact` would be added to `HashMap`, `Vec`, `String`, and `VecDeque`. These would have the exact same APIs as their infallible counterparts, except that OOM would be exposed as an error case, rather than a call to `Alloc::oom()`. They would have the following signatures:
+
+```
+/// Tries to reserve capacity for at least `additional` more elements to be inserted
+/// in the given `Vec<T>`. The collection may reserve more space to avoid
+/// frequent reallocations. After calling `reserve`, capacity will be
```

Comment:
The doc-comment should be further revised since this is the doc for `fn try_reserve`, not `fn reserve`. In particular, instead of saying "After calling `reserve`, capacity will ..." (which is not relevant to this function), you could instead say:

> If `try_reserve` returns `Ok`, capacity will be greater than or equal to `self.len() + additional`. If capacity is already sufficient, then returns `Ok` (with no side-effects to this collection).